src/environmentd/src/http.rs

```diff
@@ -187,19 +188,134 @@ pub struct InternalHttpConfig {
     pub tracing_handle: TracingHandle,
     pub adapter_client_rx: oneshot::Receiver<mz_adapter::Client>,
     pub active_connection_count: Arc<Mutex<ConnectionCounter>>,
+    pub promote_leader: mpsc::Sender<()>,
+    pub ready_to_promote: mpsc::Receiver<()>,
 }
 
 pub struct InternalHttpServer {
     router: Router,
 }
 
+#[derive(Debug, Serialize, Deserialize, PartialEq, Eq)]
+pub struct LeaderStatusResponse {
+    pub status: LeaderStatus,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub enum LeaderStatus {
+    /// The current leader returns this. It shouldn't need to know that another instance is attempting to update it.
+    IsLeader,
+    /// The new pod should return this status until it is ready to become the leader, or it has determined that it cannot proceed.
+    Initializing,
+    /// Once we receive this status, we can tell it to become the leader and migrate the EIP.
+    ReadyToPromote,
+}
+
+#[derive(Debug)]
+pub enum LeaderState {
+    IsLeader,
+    Initializing {
+        promote_leader: mpsc::Sender<()>,
+        ready_to_promote: mpsc::Receiver<()>,
+    },
+    ReadyToPromote {
+        promote_leader: mpsc::Sender<()>,
+    },
+}
+
+fn state_to_status(state: &LeaderState) -> LeaderStatus {
+    match state {
+        LeaderState::IsLeader => LeaderStatus::IsLeader,
+        LeaderState::Initializing { .. } => LeaderStatus::Initializing,
+        LeaderState::ReadyToPromote { .. } => LeaderStatus::ReadyToPromote,
+    }
+}
+
+pub async fn handle_leader_status(
+    State(state): State<Arc<Mutex<LeaderState>>>,
+) -> impl IntoResponse {
+    let mut leader_state = state.lock().expect("lock poisoned");
+    match &mut *leader_state {
+        LeaderState::IsLeader => (),
+        LeaderState::Initializing {
+            promote_leader,
+            ready_to_promote,
+        } => {
+            match ready_to_promote.try_recv() {
+                Ok(_) => {
+                    *leader_state = LeaderState::ReadyToPromote {
+                        promote_leader: promote_leader.clone(),
+                    };
+                }
+                Err(TryRecvError::Empty) => {
+                    // Continue waiting.
+                }
+                Err(TryRecvError::Disconnected) => {
+                    *leader_state = LeaderState::IsLeader; // server has started, it is the leader now
+                }
+            }
+        }
+        LeaderState::ReadyToPromote { .. } => (),
+    }
```

Comment:
Changed. Does the lint not always fire?

Reply:
The lint should fire if we ever do hold it across an await point. We're not doing that here, but dropping is a good habit to get into, in case someone else awaits something in this function in the future

Reply:
OK. I guess if it always fires then i can forget about getting it wrong (like how i lean on the compiler to enforce type safety). Added anyway

Reply:
Thanks!